docs/detective.man
.\" Automatically generated from an mdoc input file.  Do not edit.
.\" Copyright (c) 2007, 2008 Dieter Baron and Thomas Klausner.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above
.\"    copyright notice, this list of conditions and the following
.\"    disclaimer in the documentation and/or other materials provided
.\"    with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior
.\"    written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THOMAS KLAUSNER ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
.\" USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH "DETECTIVE" "1" "June 4, 2008" "NiH" "General Commands Manual"
.nh
.if n .ad l
.SH "NAME"
\fBdetective\fR
\- list zip archive contents with headers skipped
.SH "SYNOPSIS"
.HP 10n
\fBdetective\fR
[\fB\-hV\fR]
[\fB\-C\fR\ \fIhashtypes\fR]
[\fB\-D\fR\ \fIdbfile\fR]
[\fB\--db\fR\ \fIdbfile\fR]
[\fB\--detector\fR\ \fIfile\fR]
[\fB\--hash-types\fR\ \fItypes\fR]
[\fIzip-archive\fR\ [...]]
.SH "DESCRIPTION"
\fBdetective\fR
lists the contents of zip archives in a format similar to
dumpgame(1),
optionally applying a header skip detector.
.PP
If no hash types or detector are given via command line options,
they are taken from the
ckmame(1)
database.
.SS "OPTIONS"
\fBdetective\fR
supports the following options:
.TP 31n
\fB\-C\fR \fItypes\fR, \fB\--hash-types\fR \fItypes\fR
.br
A comma separated list of hash types to compute when creating a ROM
set description from a directory of zip archives.
Supported hash types are: crc, md5, sha1.
.TP 31n
\fB\-D\fR \fIdbfile\fR, \fB\--db\fR \fIdbfile\fR
Use
\fIdbfile\fR
as database file.
.TP 31n
\fB\--detector\fR \fIfile\fR
Use CMPro XML header skip detector description from
\fIfile\fR.
.SH "ENVIRONMENT"
.TP 12n
\fRMAMEDB\fR
Location for the database file.
Can be overridden with the
\fB\-D\fR
option.
Defaults to
\(lq\fImame.db\fR\(rq.
.SH "SEE ALSO"
ckmame(1),
dumpgame(1)
.SH "AUTHORS"
\fBdetective\fR
was written by
Dieter Baron \(ladillo@nih.at\(ra
and
Thomas Klausner <\fItk@giga.or.at\fR>.


NEWS
0.12 (2015-05-03)

- add support for on-disk caches to speed up consecutive runs
- support for ROMs in directories instead of in zip archives
- fix all fixable ROM errors automatically in one run
- speed up database operations by preparing statements only once
- remove torrentzip support (removed in libzip)

0.11 (2011-02-20)

- MESS Software Lists support
- retire mkmamedb-xmame.sh, use "mame -listxml | mkmamedb" instead
- link/copy disks from extra if -j (--delete-found) isn't specified
- ckmame: display current game on SIGINFO (Ctrl-T)
- removed ROMPATH support since mame uses a path in the config file nowadays

0.10 (2008-07-25)

- use SQLite3 instead of Berkeley DB
- optionally TorrentZip ROM set
- mkmamedb: add support for ROM Management Datafile format
- mkmamedb: add support to read dat files from zip archives
- mkmamedb: extract version and game description from listxml output
- ckmame: fix finding ROMs that need detectors from superfluous and extra
- change to 3-clause BSD license

0.9 (2007-06-04)

- add support for CMPro XML header skip detectors
- fix handling of zero byte ROMs
- dumpgame: brief option: no ROM/disk info
- mkmamedb: add support for reading Romcenter dat files
- mkmamedb: warn about multiple games with same name
- mkmamedb: create CMPro dat files
- mkmamedb: create mame db or CMPro dat file from zip archives
- ckmame: add option to keep files present in old ROM database

0.8 (2006-05-18)

- clean up superfluous files and extra directories
- use additional DB listing files that exist elsewhere
- read games to check from file
- create DB from multiple dat files
- omit games matching shell glob patterns from database
- don't open archives or disk images multiple times
- don't accept disk images that need a parent
- move unknown files to directory `unknown', not to `garbage' in ROM path

0.7 (2005-12-28)

- fix all fixable errors in at most two runs
- rename disks
- support searching for files in additional directories
- detect faked ROMs (correct CRC, wrong MD5/SHA1)
- adapt mkmamedb-xmame.sh for MAME version numbers of 0.100 and above
- dropped support for inferring flags/status from CRC (old MAME versions)
- deprecated -u/-U command line options
- simplify ROM matching logic
- many new regression tests, causing
- various bug fixes

0.6 (2005-06-17)

- libzip is distributed separately
- Support MAME/MESS -listxml output
- Adapt for MAME -listinfo output changes (honor flags baddump and nodump).
- Check disk images (chd files).
- Superfluous merged into ckmame. ckmame now by default reports
  extra files.  Also works for samples.
- add file format version to database and don't use incompatible database
- dumpgame: find game by checksum of one of its ROMs
- add script to create database by running xmame

0.5 (2003-01-30)

MESS, Impact, Raine, and others based on the CMPro format now supported.
Man page added.
Bug fixes.

0.4.1 (1999-10-06)

If a a clone zip didn't exist, it was not created.
The DOS port now really can fix romsets.

0.4 (1999-09-23)

The whole program.


src/detective.c
/*
  detective.c -- list files from zip archive with headers skipped
  Copyright (C) 2007-2014 Dieter Baron and Thomas Klausner

  This file is part of ckmame, a program to check rom sets for MAME.
  The authors can be contacted at <ckmame@nih.at>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions
  are met:
  1. Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in
     the documentation and/or other materials provided with the
     distribution.
  3. The name of the author may not be used to endorse or promote
     products derived from this software without specific prior
     written permission.

  THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS
  OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
  IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <stdio.h>
#include <stdlib.h>

#include "archive.h"
#include "compat.h"
#include "detector.h"
#include "error.h"
#include "globals.h"
#include "hashes.h"
#include "romdb.h"

char *usage = "Usage: %s [-hV] [-C types] [-D dbfile] [--detector detector] zip-archive [...]\n";

char help_head[] = "detective (" PACKAGE ") by Dieter Baron and"
		   " Thomas Klausner\n\n";

char help[] = "\n"
	      "  -h, --help               display this help message\n"
	      "  -V, --version            display version number\n"
	      "  -C, --hash-types types   specify hash types to compute (default: all)\n"
	      "  -D, --db dbfile          use mame-db dbfile\n"
	      "      --detector xml-file  use header detector\n"
	      "  -u, --roms-unzipped      ROMs are files on disk, not contained in zip archives\n"
	      "\n"
	      "Report bugs to " PACKAGE_BUGREPORT ".\n";

char version_string[] = "detective (" PACKAGE " " VERSION ")\n"
			"Copyright (C) 2007-2014 Dieter Baron and Thomas Klausner\n" PACKAGE " comes with ABSOLUTELY NO WARRANTY, to the extent permitted by law.\n";

#define OPTIONS "hC:DuV"

enum { OPT_DETECTOR = 256 };

struct option options[] = {
    {"help", 0, 0, 'h'}, {"version", 0, 0, 'V'}, {"db", 1, 0, 'D'}, {"detector", 1, 0, OPT_DETECTOR}, {"hash-types", 1, 0, 'C'}, {"roms-unzipped", 0, 0, 'u'}, {NULL, 0, 0, 0},
};


static int print_archive(const char *, int);
static void print_checksums(hashes_t *, int);


int
main(int argc, char **argv) {
    char *dbname;
    char *detector_name;
    int c, i, ret;
    romdb_t *ddb;
    int hashtypes;

    setprogname(argv[0]);

    detector = NULL;
    hashtypes = -1;

    dbname = getenv("MAMEDB");
    if (dbname == NULL)
	dbname = DBH_DEFAULT_DB_NAME;
    detector_name = NULL;
    roms_unzipped = 0;

    opterr = 0;
    while ((c = getopt_long(argc, argv, OPTIONS, options, 0)) != EOF) {
	switch (c) {
	case 'h':
	    fputs(help_head, stdout);
	    printf(usage, getprogname());
	    fputs(help, stdout);
	    exit(0);
	case 'V':
	    fputs(version_string, stdout);
	    exit(0);
	case 'C':
	    hashtypes = hash_types_from_str(optarg);
	    if (hashtypes == 0) {
		fprintf(stderr, "%s: illegal hash types '%s'\n", getprogname(), optarg);
		exit(1);
	    }
	    break;
	case 'D':
	    dbname = optarg;
	    break;
	case 'u':
	    roms_unzipped = 1;
	    break;
	case OPT_DETECTOR:
	    detector_name = optarg;
	    break;
	default:
	    fprintf(stderr, usage, getprogname());
	    exit(1);
	}
    }

    if (argc == optind) {
	fprintf(stderr, usage, getprogname());
	exit(1);
    }

    if (detector_name) {
	if ((detector = detector_parse(detector_name)) == NULL) {
	    myerror(ERRSTR, "cannot parse detector '%s'", detector_name);
	    exit(1);
	}
    }

    if ((ddb = romdb_open(dbname, DBH_READ)) == NULL) {
	if (detector == 0) {
	    myerror(ERRSTR, "can't open database '%s'", dbname);
	    exit(1);
	}
    }
    else {
	if (detector == NULL)
	    detector = romdb_read_detector(ddb);
	if (hashtypes == -1)
	    hashtypes = romdb_hashtypes(ddb, TYPE_ROM);
	romdb_close(ddb);
    }

    ret = 0;
    for (i = optind; i < argc; i++)
	ret |= print_archive(argv[i], hashtypes);

    return ret ? 1 : 0;
}


static int
print_archive(const char *fname, int hashtypes) {
    archive_t *a;
    file_t *f;
    int i, j, ret;

    if ((a = archive_new(fname, TYPE_ROM, FILE_NOWHERE, ARCHIVE_FL_NOCACHE)) == NULL)
	return -1;

    printf("%s\n", archive_name(a));

    ret = 0;
    for (i = 0; i < archive_num_files(a); i++) {
	if (archive_file_compute_hashes(a, i, hashtypes) < 0) {
	    ret = -1;
	    continue;
	}

	f = archive_file(a, i);

	if (file_sh_is_set(f, FILE_SH_DETECTOR))
	    j = FILE_SH_DETECTOR;
	else
	    j = FILE_SH_FULL;

	printf("\tfile %-12s  size %7" PRIu64, file_name(f), file_size_xxx(f, j));
	print_checksums(file_hashes_xxx(f, j), hashtypes);
	if (j == FILE_SH_DETECTOR)
	    printf("  (header skipped)");
	printf("\n");
    }

    return ret;
}


static void
print_checksums(hashes_t *hashes, int hashtypes) {
    int i;
    char h[HASHES_SIZE_MAX * 2 + 1];

    for (i = 1; i <= HASHES_TYPE_MAX; i <<= 1) {
	if (hashes_has_type(hashes, i) && (hashtypes & i)) {
	    printf(" %s %s", hash_type_string(i), hash_to_string(h, i, hashes));
	}
    }
}
